src/util/invariant.cpp

```diff
@@ -0,0 +1,154 @@
+// to check if the C library is glibc.
+#include <assert.h>
+#ifdef __GLIBC__
+
+// GCC needs LINKFLAGS="-rdynamic" to give function names in the backtrace
+#include <execinfo.h>
+#include <cxxabi.h>
+
+
+/*******************************************************************\
+
+Function: output_demangled_name
+
+  Inputs: The output stream and the description of the stack_entry
+
+ Outputs: True <=> the entry has been successfully demangled and printed.
+
+ Purpose: Attempts to demangle the function name assuming the glibc
+          format of stack entry:
+
+          path '(' mangled_name '+' offset ') [' address ']\0'
+
+\*******************************************************************/
+
+static bool output_demangled_name(
+  std::ostream &out,
+  const char * const stack_entry)
+{
+  bool return_value=false;
+
+  std::string working(stack_entry);
+
+  std::string::size_type start=working.rfind('(');  // Path may contain '(' !
+  std::string::size_type end=working.find('+', start);
+
+  if(start!=std::string::npos &&
+     end!=std::string::npos &&
+     start+1<=end-1)
+  {
+    std::string::size_type length=end - (start+1);
+    std::string mangled(working.substr(start+1, length));
+
+    int demangle_success=1;
+    char *demangled=
+      abi::__cxa_demangle(mangled.c_str(), NULL, 0, &demangle_success);
+
+    if(demangle_success==0)
+    {
+      out << working.substr(0, start+1)
+          << demangled
+          << working.substr(end);
+      return_value=true;
+    }
+
+    free(demangled);
+  }
+
+  return return_value;
+}
+#endif
+
+
+
+/*******************************************************************\
+
+Function: check_invariant
+
+  Inputs: The file, function and line of the invariant, plus the
+          condition to check and the reason why it is true.
+
+ Outputs: Does not return if condition is false.
+          Returns with no output or state change if true.
+
+ Purpose: Checks that the given invariant condition holds and prints a
+          back trace and / or throws an exception depending on build
+          configuration.
+
+\*******************************************************************/
+
+
+void check_invariant(const char * const file, const char * const function,
+                     const int line,
+                     const bool condition, const char * const reason)
+{
+  if(condition)
+    return;
+
+#ifdef CPROVER_INVARIANT_PRINT_STACK_TRACE
+    std::ostream & out(std::cerr);
+#else
+    std::ostringstream out;
+#endif
+
+    // Flush regularly so that errors during output will result in
+    // partial error logs rather than nothing
+    out << "Invariant check failed\n" << std::flush;
+    out << "File " << file
+        << " function " << function
+        << " line " << line
+        << '\n' << std::flush;
+
+#ifdef __GLIBC__
+    out << "Backtrace\n" << std::flush;
+
+    void * stack[50];
+    std::size_t entries;
+    char **description;
+    std::size_t i;
```

Comment:
Refactored all but stack so that the are declared and initialised together.  stack is initialised on the line below it's declaration.